documentation/design/security-metadata.md

```diff
@@ -0,0 +1,146 @@
+
+# Security Metadata
+
+The feature is meant to improve the security of builds executed via MSBuild, by reducing the chances of spilling secrets (and possibly other sensitive data) from otherwise secured or/and inaccessible build environments.
+
+It builds upon the other efforts reducing the cases accidentaly logging secrets - ['not logging unused environemnt variables'](https://github.com/dotnet/msbuild/pull/7484), 'redacting known secret patterns' (internal). Distinction here is that we want to give users option how to configure their build data so that they can indicate what contains secret/sensitive data and shouldn't get output into logs.
+
+The feature is envisioned to be facilitated via global items and/or properties that will be masking logging of specific types of log entries.
+
+Out of scope vision contains categorization of tasks (e.g. 'trustworthy'/'unknown' and 'outputing input data'/'not outputing input data'/'unknown') and passing data marked as sensitive/secrets would not be allowed (would lead to build error) based on specific configurations tunable by user. So e.g. it would not be possible to pass secrets to [`WriteLinesToFile` task](https://learn.microsoft.com/en-us/visualstudio/msbuild/writelinestofile-task).
+
+# Scope
+
+## In scope
+ * Following data can be opted-in for redacting:
+    * property values
+    * item values
+    * item metadata values
+    * all item metadata
+    * any combination of above
+    * task OutputItems (? do we want to support this as additional data type? Can be hand in cases like [`ReadLinesFromFile` task](https://learn.microsoft.com/en-us/visualstudio/msbuild/readlinesfromfile-task))
+ * Redacting the above select data from all log events (before they are being sent to loggers). Some examples of additional places where the data can propagate and hence attempted to be logged:   
+    * task input parameters
+    * task OutputItems (? do we want to support this as possible additional 'transformation' of property/item values? Possibly only when output is equal to the sensitive input) 
+    * referenced/evaluated environemnt variables
+    * input command line
+    * properties/items evalution - causing value to be transfered to a new holder (`<MyProp>$(SomeSecret)</MyProp>`)
+
+## Scope to be decided
+ * concatentaing property with other values or flattening item values or transforming items and then passing via other property - should such be ignored or redacted as whole, or redacting just the part formed from the sensitive data?
+ * values created via [Property functions](https://learn.microsoft.com/en-us/visualstudio/msbuild/property-functions) or [Item functions](https://learn.microsoft.com/en-us/visualstudio/msbuild/item-functions).
+
+## Out of scope
```

Comment:
>Out of scope [](http://example.com/codeflow?start=3&length=12)

In previous exploration, we explicitly excluded processing path information, as this is very copious in the processing stream. It is theoretically possible someone might encode a password in a source file or directory name but seems unlikely.

Reply:
Yeah - this is a good point. I can imagine someone would want to redact e.g. their login (https://github.com/dotnet/msbuild/issues/8493) - we need to find proper balance with acceptable performance (e.g. opt-in levels of redacting)